### PR TITLE
Workaround for null pointer exception in resume exercise on Jenkins setup

### LINKED ISSUE #2600: Triggering an inactive participation fails
### Describe the bug
Triggering an inactive participation fails with NPE.

#### To Reproduce
Go to an inactive participation. Press the red "trigger build" button. It fails with NPE in the logs and nothing happens.

#### Expected behavior
It should not fail. Reason for the failure is that in an inactive participation the build plan id is null. But the line 
ContinuousIntegrationService.BuildStatus buildStatus = continuousIntegrationService.get().getBuildStatus(programmingExerciseParticipation);
in ProgrammingSubmissionRessource.java
tries to look up the build plan on the server and fails when trying to compose the folder name.

### Environment
<details><pre>
 - OS: Linux server
 - Browser [e.g. chrome, safari]: Firefox
 - Artemis version: 4.8.4
 - Jenkins/Gitlab connectors
</pre></details>

#### Additional context

Suggested bugfix: Do not ask the CI service whether a build is ongoing if the participation is inactive.

### PR DESCRIPTION
**Pull request by krusche**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.

### Motivation and Context
Fixes #2600 and #2601 

### Description
Implements a workaround and adds TODOs for future improvements
I also tried to improve the JavaDoc and method parameter names

### Steps for Testing
1. Create and participate in a programming exercise on TS2 (Gitlab / Jenkins) 
2. Cleanup the build plan (not the repository) several times and try out the following methods
3. Resume the exercise (as student) by clicking on resume exercise
4. Resume the exercise by pushing into the repo (the build should automatically get started afterwards and you should still receive a result)
5. Trigger a build as instructor (should automatically resume the exercise)
6. Trigger all builds as instructor (should automatically resume the exercise for multiple users)

**Important:** please review that I did not mix up anything while renaming parameters and variables

### Test Coverage
no changes

### Screenshots
no UI changes

### CODE REVIEW

**Review by airelawaleria — CHANGES_REQUESTED** [review]
Tested in the testing session on ts2. Could not have resumed the exercise from Artemis, got an error. Also could not have pushed to the repository. @JohannesWeiss @simonzilker @LudwigWeide 
image
image

**Review by simonzilker — COMMENTED** [review]
Tested in testing session. Faced same issue as @airelawaleria

**Review by JohannesWeiss — COMMENTED** [review]
Tested in testing session, can confirm issue mentioned by @airelawaleria.

**Review by LudwigWeide — CHANGES_REQUESTED** [review]
Tested with @airelawaleria . Same issue.

**Review by marvinchimeka — COMMENTED** [review]
Faced the same issue.

**Review by tiffca — CHANGES_REQUESTED** [review]
I faced several issues while testing. 
1. ISE when resuming the exercise 
2. can't resume the exercise through pushing a change 
3. can't resume the build as an instructor/student after cleanup 

ISE when resuming the exercise
Steps to reproduce: 
1. create a programming exercise 
2. as a student make a change and submit
3. as an instructor go to exercise participations and click cleanup 
4. as a student, click on resume the exercise. 
5. Result: you get an ISE. Expected: no ISE and it resumes the exercise 
6. You can click resume again and it will open the code editor 

Error log:  HRPUThttps://artemistest2.ase.in.tum.de/api/courses/1439/exercises/2455/resume-programming-participation
[HTTP/2 500 Internal Server Error 779ms]
image

 can't resume the exercise through pushing a change/can't resume the build as an instructor/student after cleanup  
steps: 
1. create a programming exercise 
2. as a student, clone the repo and push a change, e.g. implement bubble sort
3.  as an instructor go to exercise participations and click cleanup 
4. as a student push a change to try to resume the exercise. For example implement merge sort. 
5. Result: it goes to "Start exercise" instead of "Resume". The build doesn't reflect the latest pushed changes. 
6. as a student make a change, e.g. remove merge sort
6. As an instructor, try to cleanup, trigger the build again, trigger all. Result: the build does not finish building or does not reflect the latest changes. As a student you receive this message "The result for your last submission is not available. Please submit your code again."
7. as a student, try to make a change and trigger the build. Result: it doesn't reflect the latest changes. e.g. showing 2 passes when only 1 is implemented.

**Comment by krusche** [comment]
Thanks for testing, I'll look into the issue and try to fix it

**Comment by krusche** [comment]
I think the issue with the repository found by @airelawaleria is unrelated. When you cleanup the exercise, you need to make sure that only the build plan is deleted, but not the repository

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java [hunk 1]
```diff
@@ -450,7 +450,16 @@ private void setupTestTemplateAndPush(Repository repository, Resource[] resource
 
                     try {
                         Resource[] projectTypeTestFileResources = resourceLoaderService.getResources(projectTypeTemplatePath);
```
Comment (ivanchimeno) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java:452]: If we add a method `resourceLoaderService.getExistantResources(projectTypeTemplatePath)` then we can move the lines written bellow into that method and save some space here

**Review by ivanchimeno — DISMISSED** [review]
Tested on ts2 and seems to be working fine

**Comment by krusche** [comment]
I added more tests for the resume case, there are still some missing (see TODOs in the test code), but I think we can add those in a follow-up PR.

**Review by airelawaleria — APPROVED** [review]
Tested with @simonzilker. Everything works fine, but the resume button does not appear in the course's student view at first, but only after the exercise view (with the prob. statement) is opened once. Probably it's expected behavior, so approve.

**Review by simonzilker — APPROVED** [review]
Tested with @airelawaleria. Everything works fine, but the issue by @airelawaleria. See video below.

https://user-images.githubusercontent.com/73840817/103268080-2b569400-49b3-11eb-91dd-19a95cf20dd7.mp4

**Review by JohannesWeiss — APPROVED** [review]
Tested on ts2 as well and worked as expected for all ways of resuming, aside of the aforementioned issue with the resume button.

**Review by marvinchimeka — APPROVED** [review]
Tested on ts2. Works as intended. There is an issue with the resume button which was already mentioned by reviewers above.

**Review by tiffca — APPROVED** [review]
The push issue is resolved and works as expected. 

I still notice a discrepancy with instructor trigger build or trigger all. But I guess it's as expected behavior for the others, hence approving. 

Steps to reproduce: 
1. As a student participate in an exercise, implement merge sort and bubble sort and submit
2. as an instructor clean up the build 
3. as a student remove merge sort implementation and save
4. as an instructor trigger the build 
5. Expected: only 1 should be implemented, Actual: it still shows 2 implemented

**Comment by krusche** [comment]
> The push issue is resolved and works as expected.
> 
> I still notice a discrepancy with instructor trigger build or trigger all. But I guess it's as expected behavior for the others, hence approving.
> 
> Steps to reproduce:
> 
> 1. As a student participate in an exercise, implement merge sort and bubble sort and submit
> 2. as an instructor clean up the build
> 3. as a student remove merge sort implementation and save
> 4. as an instructor trigger the build
> 5. Expected: only 1 should be implemented, Actual: it still shows 2 implemented
> 6. Workaround: it works if the student then submits their code

Thanks for the detailed description: I guess the issue is in step 3: When the student only saves the changes in the online editor, they are not yet submitted into the repo.
When resuming the exercise, an empty commit has to be performed, which takes the saved state in the online code editor. This might be implemented differently for instructor trigger build and trigger all.

However, please notice, that this is an absolute edge case. Build plans are only cleaned after a student is successful or after some longer time of inactivity, so it is very unlikely that there are still changes only saved and the instructor then triggers the build manually.

However, we can try to fix this in a follow-up PR.

**Comment by krusche** [comment]
@tiffca: I just looked it up in the code. There should be no notable difference with saved vs. submitted code in the online code editor for trigger one vs. trigger all. The empty commit only will be executed when the student clicks the "resume exercise" button. Then this commit would include the last saved state of the online code editor.

If the student does not click the resume exercise button, then not commit / push is executed from the online code editor.
While trigger one and trigger all have different starting code, they both end in the same function that resumes the exercise for the student(s) if necessary and triggers one build on the CI server.

**Comment by krusche** [comment]
Ah, I just re-read your message. The difference was between trigger build and the student clicking on resume exercise. Yes, this would then be expected behavior.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java (hunk 1 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java:452]: Good catch. Thanks!

**Comment by krusche** [comment]
> Tested with @airelawaleria. Everything works fine, but the issue by @airelawaleria. See video below.
> 
>  Course.details.-.Google.Chrome.2020-12-29.08-50-09.mp4

This is ok, the issue here is that the browser page was not refreshed after the manual cleanup by the instructor. And the cleanup is not communicated to the student automatically (also the student could be offline). The good thing is that the cleanup does not happen often and navigating into the exercise reloads the exercise details, so I think we can live with this.

**Pull request merged**